### PR TITLE
[CONTACT-928] : Multiple emails text input

### PR DESCRIPTION
**Pull request by Moniet**
Text Input with ability to add multiple emails. 

Emails are created when keyboard events 'Enter', 'Space', or 'Comma' occurs. This functionality is abstracted to the `useMultipleEmailsInput` hook.

Screenshot 2021-12-22 at 7 09 41 PM

### CODE REVIEW

File: src/components/MultipleEmailsInput/MultipleEmailsInput.tsx [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+    placeholder,
+    hasError,
+    message,
+    value,
+    deleteEmail,
+    className,
+}: MultipleEmailsInputProps) => {
+    return (
+        <MultipleEmailsInputContainer onKeyDown={(e) => onKeyDown(e)} {...(className ? { className } : {})}>
+            <InputUsageDirections hasError={hasError}>{INPUT_USAGE_DIRECTIONS}</InputUsageDirections>
```
Comment (jlambdev) [inline src/components/MultipleEmailsInput/MultipleEmailsInput.tsx:27]: Can we allow this to be optional / pass a string value for the usage directions? A few things come to mind:

- Supporting different locales (parent component can be responsible for providing this value)
- There will be cases where we want to control if / how the guidance text would be displayed

File: src/components/stories/MultipleEmailsInput.stories.tsx [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+    title: 'Components/MultipleEmailsInput',
+};
+
+export const ExampleWithoutEmails = () => {
+    const { inputProps, emails } = useMultipleEmailsTextInput();
+    const persistEmails = (allEmails: string[]) => allEmails;
+
+    useEffect(() => {
+        persistEmails(emails);
+    }, [emails]);
```
Comment (jlambdev) [inline src/components/stories/MultipleEmailsInput.stories.tsx:14]: Nice demonstration within the context of a story.

**Review by jlambdev — COMMENTED** [review]
Very nice:  a few comments & considerations.

**Comment by jlambdev** [comment]
Also just for other devs reference: RTL tests are planned, but would be covered in a separate ticket

File: src/components/MultipleEmailsInput/useMultipleEmailsInput.ts [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+        });
+
+        return Object.keys(emailsMap);
+    };
+
+    const getEscapedEmailValue = (value = '') => value.replace(/,|\s/, ''); // escapes ',' (commas) and ' ' (i.e. any whitespace character)
+
+    const showErrorMessage = () => {
+        setHasError(true);
+        setMessage('Email structure must be x@y.z');
```
[the remark this answers is not in the document]
Comment (Moniet) [reply src/components/MultipleEmailsInput/useMultipleEmailsInput.ts:46]: Very good point !

[the remark this answers is not in the document]
Comment (Moniet) [reply src/components/MultipleEmailsInput/useMultipleEmailsInput.ts:46]: Have added an optional prop here.

File: src/components/MultipleEmailsInput/useMultipleEmailsInput.ts [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+
+    const onChange = ({ value }: { value: string }) => setInputValue(value);
+
+    const onKeyDown = (e: KeyboardEvent<HTMLDivElement>) => {
+        // keyboard event bubbles up from the text input :
+        // we handle 'Enter' | ',' | ' ' as actions to update the email list
+
+        const email = getEscapedEmailValue(inputValue);
+        const isEmailValid = isValidEmailAddress(email);
+        const isCreateEmailAction = shouldCreateEmail(e?.key);
```
[the remark this answers is not in the document]
Comment (Moniet) [reply src/components/MultipleEmailsInput/useMultipleEmailsInput.ts:62]: I agree, this makes sense for readability. Have switched it up as suggested.

File: src/components/MultipleEmailsInput/MultipleEmailsInput.tsx (hunk 1 above)
Comment (Moniet) [reply src/components/MultipleEmailsInput/MultipleEmailsInput.tsx:27]: Makes sense, have added necessary prop for this.

**Review by jlambdev — APPROVED** [review]

File: src/components/stories/MultipleEmailsInput.stories.tsx [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+                {...inputProps}
+                name="multipleEmailsInput"
+                message="Custom Error Message Here" // overridding inputProps
+                hasError // overridding inputProps
+            />
+        </div>
+    );
+};
+
+export const i18nErrorMessages = () => {
```
Comment (harman052) [inline src/components/stories/MultipleEmailsInput.stories.tsx:72]: Looks good, one minor comment. Name of this example appears in storyBook as: "I 18 N Error Messages". You can improve it you like to.

File: src/components/MultipleEmailsInput/styled.ts [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+    width: fit-content;
+    padding: ${({ theme: { spacing } }) => `${spacing(1)} ${spacing(3)}`};
+    background-color: ${({ theme: { palettes } }) => palettes.greyScale.main.bgColor};
+    border-radius: 50px;
+    color: ${({ theme: { palettes } }) => palettes.info.light.contrast};
+`;
+
+export const EmailChipsContainer = styled.div`
+    display: flex;
+    flex-direction: column;
```
Comment (harman052) [inline src/components/MultipleEmailsInput/styled.ts:14]: Is it desired that the chips should appear as a column?

File: src/components/stories/MultipleEmailsInput.stories.tsx (hunk 5 above)
Comment (Moniet) [reply src/components/stories/MultipleEmailsInput.stories.tsx:72]: Good catch, will change this !

File: src/components/MultipleEmailsInput/styled.ts (hunk 6 above)
Comment (Moniet) [reply src/components/MultipleEmailsInput/styled.ts:14]: According to the Figma file / discussions, there wouldn't be a need for more than 2-3 emails at max. 
So it was decided to keep this as a column. 

My thought pattern was to export each element here separately so devs can compose them as needed.

**Review by harman052 — APPROVED** [review]

**Pull request merged**